modules/internal/ChapelRange.chpl

```diff
@@ -1454,21 +1454,67 @@ proc _cast(type t, r: range(?)) where isRangeType(t) {
     compilerError("can't apply 'by' to a range with idxType ",
                   int(w):string, " using a step of type ",
                   stride.type:string);
-    yield nil; // iters needs a yield in them
+    yield nil;
   }
 
   iter chpl_direct_range_iter(low: uint(?w), high: uint(w), stride) {
     compilerError("can't apply 'by' to a range with idxType ",
                   uint(w):string, " using a step of type ",
                   stride.type:string);
-    yield nil; // iters needs a yield in them
+    yield nil;
   }
 
-
   // case for when low and high aren't compatible types and can't be coerced
   iter chpl_direct_range_iter(low, high, stride) {
-    compilerError("Bounds of '..' must be integers of compatible types, when specified.");
-    yield nil; // iters needs a yield in them
+    compilerError("Bounds of 'low..high' must be integers of compatible types.");
+    yield nil;
+  }
+
+
+  //
+  // Direct range iterators for low bounded counted ranges (low..#count)
+  //
+
+  iter chpl_direct_counted_range_iter(low: int(?w), count: int(w)) {
+    for i in chpl_direct_real_counted_range_iter(low, count) do yield i;
+  }
+
+  iter chpl_direct_counted_range_iter(low: int(?w), count: uint(w)) {
+    for i in chpl_direct_real_counted_range_iter(low, count) do yield i;
+  }
+
+  iter chpl_direct_counted_range_iter(low: uint(?w), count: int(w)) {
+    for i in chpl_direct_real_counted_range_iter(low, count) do yield i;
+  }
+
+  iter chpl_direct_counted_range_iter(low: uint(?w), count: uint(w)) {
+    for i in chpl_direct_real_counted_range_iter(low, count) do yield i;
+  }
+
+  iter chpl_direct_counted_range_iter(low: integral, count) {
+    compilerError("can't apply '#' to a range with idxType ",
+                  low.type:string, " using a count of type ",
+                  count.type:string);
+    yield nil;
+  }
+
+  iter chpl_direct_counted_range_iter(low, count) {
+    compilerError("Bound of 'low..' must be an integer");
+    yield nil;
+  }
+
+  // The "actual" counted range iter. Turn the bounds of a low bounded counted
+  // range into the bounds of a fully bounded non-strided range. `low..#count`
+  // becomes `low..(low + (count - 1))`. Needs to check for negative counts,
+  // and for zero counts iterates over a degenerate `1..0`.
+  iter chpl_direct_real_counted_range_iter(low, count) {
```

Comment:
The use of `real` here suggests "real floating point value" to me rather than "I do the real work."  Maybe "helper"?  (in which case I'd probably move it to the end of the routine's name: `chpl_direct_counted_range_iter_helper`